// treat/lib/commonjs/webpackTreat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var scopeCount = 0;
var mockThemes = [];
var webpackTreat = {
    addLocalCss: function () { },
    addThemedCss: function () { },
    addTheme: function (theme) { return mockThemes.push(theme); },
    getThemes: function () { return mockThemes; },
    getIdentName: function (local, _, theme) { return "" + (theme ? '_' : '') + local; },
};
exports.setWebpackTreat = function (newWebpackTreat) {
    scopeCount = 0;
    webpackTreat = newWebpackTreat;
};
exports.addLocalCss = function () {
    var props = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        props[_i] = arguments[_i];
    }
    return webpackTreat.addLocalCss.apply(webpackTreat, props);
};
exports.addThemedCss = function () {
    var props = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        props[_i] = arguments[_i];
    }
    return webpackTreat.addThemedCss.apply(webpackTreat, props);
};
exports.addTheme = function () {
    var props = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        props[_i] = arguments[_i];
    }
    return webpackTreat.addTheme.apply(webpackTreat, props);
};
exports.getThemes = function () {
    var props = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        props[_i] = arguments[_i];
    }
    return webpackTreat.getThemes.apply(webpackTreat, props);
};
exports.getIdentName = function () {
    var props = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        props[_i] = arguments[_i];
    }
    return webpackTreat.getIdentName.apply(webpackTreat, props);
};
exports.getNextScope = function () { return scopeCount++; };
